Compile bug: Unable to build on OpenBSD 7.6

### Git commit

f66f5829276650cd83a087ab2cfed1a760183ea1

### Operating systems

BSD

### GGML backends

CPU

### Problem description & steps to reproduce

I am trying to compile on OpenBSD 7.6 (GENERIC.MP#338 amd64). 

`cmake -B build ` completes with a warning.

`cmake --build build --config Release` fails with 2 errors.

I'm able to build successfully on `Linux 6.12.7-arch1-1`, following the same steps.

### First Bad Commit

_No response_

### Relevant log output

```shell
$ cmake -B build   
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: amd64
-- Including CPU backend
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND) 
CMake Warning at ggml/src/ggml-cpu/CMakeLists.txt:53 (message):
  OpenMP not found
Call Stack (most recent call first):
  ggml/src/CMakeLists.txt:299 (ggml_add_cpu_backend_variant_impl)


-- x86 detected
-- Adding CPU backend variant ggml-cpu: -march=native 
-- Configuring done (0.6s)
-- Generating done (0.9s)
-- Build files have been written to: /home/dwrz/llama.cpp/build

$ cmake --build build --config Release                    
[  4%] Built target ggml-base
[  9%] Built target ggml-cpu
[ 10%] Built target ggml
[ 10%] Building CXX object src/CMakeFiles/llama.dir/llama.cpp.o
In file included from /home/dwrz/llama.cpp/src/llama.cpp:4:
/home/dwrz/llama.cpp/src/./llama-mmap.h:21:9: error: expected member name or ';' after declaration specifiers
    int fileno() const;
    ~~~ ^
/usr/include/stdio.h:414:20: note: expanded from macro 'fileno'
#define fileno(p)       (!__isthreaded ? __sfileno(p) : (fileno)(p))
                         ^
In file included from /home/dwrz/llama.cpp/src/llama.cpp:4:
/home/dwrz/llama.cpp/src/./llama-mmap.h:21:9: error: expected ')'
/usr/include/stdio.h:414:20: note: expanded from macro 'fileno'
#define fileno(p)       (!__isthreaded ? __sfileno(p) : (fileno)(p))
                         ^
/home/dwrz/llama.cpp/src/./llama-mmap.h:21:9: note: to match this '('
/usr/include/stdio.h:414:19: note: expanded from macro 'fileno'
#define fileno(p)       (!__isthreaded ? __sfileno(p) : (fileno)(p))
                        ^
2 errors generated.
*** Error 1 in . (src/CMakeFiles/llama.dir/build.make:76 'src/CMakeFiles/llama.dir/llama.cpp.o': cd /home/dwrz/llama.cpp/build/src && /usr/b...)
*** Error 2 in . (CMakeFiles/Makefile2:1767 'src/CMakeFiles/llama.dir/all': /usr/bin/make -s -f src/CMakeFiles/llama.dir/build.make src/CMak...)
*** Error 2 in /home/dwrz/llama.cpp/build (Makefile:146 'all': /usr/bin/make -s -f CMakeFiles/Makefile2 all)
```

Unfortunately not! It makes some progress, but then encounters an error at `llama-mmap.cpp:248:14`.

```
$ cmake -B build
-- The C compiler identification is Clang 16.0.6
-- The CXX compiler identification is Clang 16.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: /usr/local/bin/git (found version "2.46.1")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
-- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
-- CMAKE_SYSTEM_PROCESSOR: amd64
-- Including CPU backend
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND) 
CMake Warning at ggml/src/ggml-cpu/CMakeLists.txt:53 (message):
  OpenMP not found
Call Stack (most recent call first):
  ggml/src/CMakeLists.txt:299 (ggml_add_cpu_backend_variant_impl)


-- x86 detected
-- Adding CPU backend variant ggml-cpu: -march=native 
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Configuring done (2.3s)
-- Generating done (1.0s)
-- Build files have been written to: /home/dwrz/llama.cpp/build

$ cmake --build build --config Release
[  1%] Building C object ggml/src/CMakeFiles/ggml-base.dir/ggml.c.o
[  1%] Building C object ggml/src/CMakeFiles/ggml-base.dir/ggml-alloc.c.o
[  2%] Building CXX object ggml/src/CMakeFiles/ggml-base.dir/ggml-backend.cpp.o
[  2%] Building CXX object ggml/src/CMakeFiles/ggml-base.dir/ggml-opt.cpp.o
[  3%] Building CXX object ggml/src/CMakeFiles/ggml-base.dir/ggml-threading.cpp.o
[  3%] Building C object ggml/src/CMakeFiles/ggml-base.dir/ggml-quants.c.o
[  4%] Linking CXX shared library libggml-base.so
[  4%] Built target ggml-base
[  4%] Building C object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu.c.o
[  5%] Building CXX object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu.cpp.o
[  5%] Building CXX object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu-aarch64.cpp.o
[  6%] Building CXX object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu-hbm.cpp.o
[  6%] Building C object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu-quants.c.o
[  7%] Building CXX object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu-traits.cpp.o
[  7%] Building CXX object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/amx/amx.cpp.o
[  8%] Building CXX object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/amx/mmq.cpp.o
[  8%] Building CXX object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/llamafile/sgemm.cpp.o
[  9%] Linking CXX shared library libggml-cpu.so
[  9%] Built target ggml-cpu
[ 10%] Building CXX object ggml/src/CMakeFiles/ggml.dir/ggml-backend-reg.cpp.o
[ 10%] Linking CXX shared library libggml.so
[ 10%] Built target ggml
[ 10%] Building CXX object src/CMakeFiles/llama.dir/llama.cpp.o
[ 11%] Building CXX object src/CMakeFiles/llama.dir/llama-adapter.cpp.o
[ 11%] Building CXX object src/CMakeFiles/llama.dir/llama-arch.cpp.o
[ 12%] Building CXX object src/CMakeFiles/llama.dir/llama-batch.cpp.o
[ 12%] Building CXX object src/CMakeFiles/llama.dir/llama-chat.cpp.o
[ 13%] Building CXX object src/CMakeFiles/llama.dir/llama-context.cpp.o
[ 13%] Building CXX object src/CMakeFiles/llama.dir/llama-grammar.cpp.o
[ 14%] Building CXX object src/CMakeFiles/llama.dir/llama-hparams.cpp.o
[ 14%] Building CXX object src/CMakeFiles/llama.dir/llama-impl.cpp.o
[ 15%] Building CXX object src/CMakeFiles/llama.dir/llama-kv-cache.cpp.o
[ 15%] Building CXX object src/CMakeFiles/llama.dir/llama-mmap.cpp.o
/home/dwrz/llama.cpp/src/llama-mmap.cpp:248:14: error: expected unqualified-id
    return ::fileno(pimpl->fp);
             ^
/usr/include/stdio.h:414:19: note: expanded from macro 'fileno'
#define fileno(p)       (!__isthreaded ? __sfileno(p) : (fileno)(p))
                        ^
1 error generated.
*** Error 1 in . (src/CMakeFiles/llama.dir/build.make:216 'src/CMakeFiles/llama.dir/llama-mmap.cpp.o': cd /home/dwrz/llama.cpp/build/src && ...)
*** Error 2 in . (CMakeFiles/Makefile2:1767 'src/CMakeFiles/llama.dir/all': /usr/bin/make -s -f src/CMakeFiles/llama.dir/build.make src/CMak...)
*** Error 2 in /home/dwrz/llama.cpp/build (Makefile:146 'all': /usr/bin/make -s -f CMakeFiles/Makefile2 all)

$ git show HEAD
commit c98eb635d60d99633e17b36c6442487d392ceaf7 # ...
```

--- a/src/llama-mmap.cpp
+++ b/src/llama-mmap.cpp
@@ -245,7 +245,11 @@ int llama_file::file_id() const {
 #ifdef _WIN32
     return _fileno(pimpl->fp);
 #else
+#if defined(fileno)
+    return fileno(pimpl->fp);
+#else
     return ::fileno(pimpl->fp);
+#endif
 #endif
 }
 

--- a/src/llama-mmap.cpp
+++ b/src/llama-mmap.cpp
@@ -241,7 +241,7 @@ llama_file::~llama_file() = default;
 size_t llama_file::tell() const { return pimpl->tell(); }
 size_t llama_file::size() const { return pimpl->size; }
 
-int llama_file::fileno() const {
+int llama_file::file_id() const {
 #ifdef _WIN32
     return _fileno(pimpl->fp);
 #else
@@ -265,7 +265,7 @@ struct llama_mmap::impl {
 
     impl(struct llama_file * file, size_t prefetch, bool numa) {
         size = file->size();
-        int fd = file->fileno();
+        int fd = file->file_id();
         int flags = MAP_SHARED;
         if (numa) { prefetch = 0; }
 #ifdef __linux__
@@ -357,7 +357,7 @@ struct llama_mmap::impl {
 
         size = file->size();
 
-        HANDLE hFile = (HANDLE) _get_osfhandle(file->fileno());
+        HANDLE hFile = (HANDLE) _get_osfhandle(file->file_id());
 
         HANDLE hMapping = CreateFileMappingA(hFile, NULL, PAGE_READONLY, 0, 0, NULL);
 

--- a/src/llama-mmap.h
+++ b/src/llama-mmap.h
@@ -18,7 +18,7 @@ struct llama_file {
     size_t tell() const;
     size_t size() const;
 
-    int fileno() const;
+    int file_id() const; // fileno overload
 
     void seek(size_t offset, int whence) const;
 

